SANC-50-Handle-password-reset-on-login-page

wesleylui opened the pull request:
Implement password reset feature for users who've forgotten their password.
Created: 
- ResetPasswordEmailTemplate (similar to RegisterAccountEmailTemplate, but I think this file could be improved)
- the Resend call for handling normal password reset in `auth.ts`
- `/forgot-password` page
- `/reset-password` page

TODO:
- Test the full flow of email sending and password resetting
- Make password stricter
- Password strength indicator? (optional)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Forgot Password page to request a reset via email.
  * Reset Password page to set a new password using a secure token.
  * New HTML reset-password email template with personalized greeting, reset button and plain URL, and 1-hour expiry notice.

* **Improvements**
  * Centralized email validation applied across forms.
  * Slightly narrower auth layout for tighter form layout.
  * Reused shared auth styling across related pages.

* **API**
  * New endpoint to trigger sending password reset emails.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wesleylui committed:
feature: add password reset email template

wesleylui committed:
feature: implement password reset email sending

wesleylui committed:
feature: add forgot-password page

wesleylui committed:
feature: add reset-password page

burtonjong commented on src/app/forgot-password/page.tsx:
```diff
@@ -0,0 +1,105 @@
+
+export default function ForgotPasswordPage() {
+  const [loading, setLoading] = useState(false);
+  const [isSuccess, setIsSuccess] = useState(false);
```
You won't need these if you instead make this a trpc request. You would be able to access them directly from the mutation. I did something similar when actually sending the invite to the user in the admin dashboard page, so you can copy that pattern

burtonjong commented on src/app/forgot-password/page.tsx:
```diff
@@ -0,0 +1,105 @@
+      email: "",
+    },
+    validate: {
+      email: (value) => (/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(value) ? null : "Invalid email"),
```
I think we validate emails in multiple places. We should put this regex so that it can be re-used

you can put this in `src/validation.ts`. also replace any email validation with that regex

burtonjong commented on src/app/reset-password/page.tsx:
```diff
@@ -0,0 +1,130 @@
+      confirmPassword: "",
+    },
+    validate: {
+      password: (value) => (value.length > 0 ? null : "Password is required"),
```
I think we want password to be greater than 8 chars.... basically copy the validation from the what we put for the login page

promatty replied:
also describe the requirements, like "Must be at least 8 characters with uppercase, lowercase, and numbers" from the complete-registration page

burtonjong commented on src/app/reset-password/page.tsx:
```diff
@@ -0,0 +1,130 @@
+
+export default function ResetPasswordPage() {
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState("");
```
same as above. We should be using tanstack here to access loading/error/success states

burtonjong commented on src/app/reset-password/page.tsx:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  return (
+    <div style={{ maxWidth: "400px", margin: "4rem auto", padding: "0 1rem" }}>
```
These should be classes, not inline styles

promatty replied:
ya replace all instances of them pls

burtonjong commented on src/lib/auth.ts:
```diff
@@ -59,6 +60,25 @@ export const auth = betterAuth({
+        if (!resend) {
+          console.warn("Resend API key not configured, skipping email");
+          return;
+        }
```
we can just move this out of the if statement so we don't dupliacte it above

wesleylui committed:
fix: use login's password validation

wesleylui committed:
created auth-layout.module.scss styling for all auth pages (login, forgot-password, complete-registration)

wesleylui committed:
add files for last commit my bad

wesleylui committed:
feat: add emailRegex to validation.ts and apply to \login and \forgot-password

wesleylui committed:
feat: create sendPasswordResetEmail trpc request. it uses api.requestPasswordReset not .forgetPassword

promatty commented on src/app/forgot-password/page.tsx:
```diff
@@ -0,0 +1,105 @@
+    setLoading(true);
+
+    try {
+      await authClient.forgetPassword({
```
Did you mean 'resetPassword'?

wesleylui replied:
I refactored forgot-password to use the tRPC request.

in `organizations.ts` the BetterAuth APIs being called are:
- resetPassword calls resetPassword
- and sendPasswordResetEmail calls requestPasswordReset

Are those the right APIs?

promatty replied:
yes 😄

wesleylui committed:
refactor: migrate reset-password to use tRPC and Tansack Query

wesleylui commented:
Great suggestions guys, thank you.
Quick summary:
- Made password validations use passwordSchema from `validation.ts`
- I created `auth-layout.module.scss` as styling for all auth pages (login, forgot-password, reset-password, complete-registration)

And I refactored both `forgot-password.tsx` `reset-password.tsx` to use tRPC requests and Tanstack queries to use the mutation for loading, error, and success states.
This means I added a `sendPasswordResetEmail` prodecure to `organizations.ts`

burtonjong commented:
> Great suggestions guys, thank you. Quick summary:
> 
> * Made password validations use passwordSchema from `validation.ts`
> * I created `auth-layout.module.scss` as styling for all auth pages (login, forgot-password, reset-password, complete-registration)
> 
> And I refactored both `forgot-password.tsx` `reset-password.tsx` to use tRPC requests and Tanstack queries to use the mutation for loading, error, and success states. This means I added a `sendPasswordResetEmail` prodecure to `organizations.ts`

I've confirmed on my own local that this works fine. However, you still need to make the build pass before I approve

promatty commented on src/app/reset-password/page.tsx:
```diff
@@ -0,0 +1,122 @@
+    );
+  }
+
+  return (
```
you have a build error. to fix it, wrap the component in a suspense like how it's done in the complete-registration page

promatty reviewed (changes requested):
fix build

wesleylui committed:
Merge branch 'main' of https://github.com/Code-the-Change-YYC/salvationarmy into SANC-50-Handle-password-reset-on-login-page

wesleylui committed:
fix: wrap reset-password page in Suspense

promatty reviewed (approved):
lgtm
